File: native-windows/FitGenerator.Native/Core/NativeCoreService.cs
using System.Globalization;
using System.Text.Json;

namespace FitGenerator.Native.Core;

internal sealed record ActivityRoutePoint(double Latitude, double Longitude);

internal sealed record ActivityInput(
    DateTime StartTime,
    IReadOnlyList<ActivityRoutePoint> RoutePoints,
    double PaceSecondsPerKilometer,
    int RestingHeartRate,
    int MaximumHeartRate,
    int LapCount);

internal interface INativeCoreService
{
    ActivityModelDto Preview(ActivityInput input, ulong seed);

    byte[] GenerateFit(ActivityInput input, ulong seed, int variantIndex);
}

internal sealed class NativeCoreService : INativeCoreService
{
    private const uint SupportedApiVersion = 1;
    private const int SupportedSchemaVersion = 1;
    private const int SupportedAlgorithmVersion = 1;
    private const int PreviewVariantIndex = 1;
    private const int InvalidInputCode = 100;
    private const int IncompatibleVersionCode = 101;
    private const int InternalErrorCode = 900;
    private const string RouteMode = "close_if_needed";

    private readonly INativeCoreApi _api;

    internal NativeCoreService()
        : this(new NativeCoreApi())
    {
    }

    internal NativeCoreService(INativeCoreApi api)
    {
        ArgumentNullException.ThrowIfNull(api);

        _api = api;
        var apiVersion = InvokeNative(
            () => api.ApiVersion,
            "无法读取 Rust 核心 API 版本");
        if (apiVersion != SupportedApiVersion)
        {
            throw new NativeCoreException(
                IncompatibleVersionCode,
                $"Rust 核心 API 版本不兼容：需要 {SupportedApiVersion}，实际 {apiVersion}");
        }
    }

    public ActivityModelDto Preview(ActivityInput input, ulong seed)
    {
        var request = SerializeRequest(input, seed, PreviewVariantIndex);
        var payload = InvokeNative(
            () => _api.Preview(request),
            "无法调用 Rust 核心生成活动预览");

        return DeserializePreview(payload, seed);
    }

    public byte[] GenerateFit(ActivityInput input, ulong seed, int variantIndex)
    {
        var request = SerializeRequest(input, seed, variantIndex);
        var payload = InvokeNative(
            () => _api.GenerateFit(request),
            "无法调用 Rust 核心生成 FIT 文件");

        if (!HasFitSignature(payload))
        {
            throw InvalidNativePayload("Rust 核心返回的 FIT 数据无效");
        }

        return payload;
    }

    private static byte[] SerializeRequest(ActivityInput input, ulong seed, int variantIndex)
    {
        if (input is null)
        {
            throw new NativeCoreException(InvalidInputCode, "活动参数不能为空");
        }

        if (input.RoutePoints is null)
        {
            throw new NativeCoreException(InvalidInputCode, "活动路线不能为空");
        }

        var points = new GeoPointDto[input.RoutePoints.Count];
        for (var index = 0; index < input.RoutePoints.Count; index++)
        {
            var point = input.RoutePoints[index];
            if (point is null)
            {
                throw new NativeCoreException(InvalidInputCode, "活动路线包含空坐标");
            }

            points[index] = new GeoPointDto(point.Latitude, point.Longitude);
        }

        var request = new CoreRequestDto(
            SupportedSchemaVersion,
            ToUtcString(input.StartTime),
            points,
            input.PaceSecondsPerKilometer,
            input.RestingHeartRate,
            input.MaximumHeartRate,
            input.LapCount,
            variantIndex,
            seed,
            RouteMode);

        try
        {
            return JsonSerializer.SerializeToUtf8Bytes(request, CoreJson.SerializerOptions);
        }
        catch (Exception exception) when (exception is JsonException or NotSupportedException or ArgumentException)
        {
            throw new NativeCoreException(InvalidInputCode, "活动参数无法转换为 Rust 核心请求");
        }
    }

    private static ActivityModelDto DeserializePreview(byte[] payload, ulong expectedSeed)
    {
        if (payload is null || payload.Length == 0)
        {
            throw InvalidNativePayload("Rust 核心返回的活动预览数据为空");
        }

        ActivityModelDto? model;
        try
        {
            model = JsonSerializer.Deserialize<ActivityModelDto>(
                payload,
                CoreJson.SerializerOptions);
        }
        catch (Exception exception) when (exception is JsonException or NotSupportedException)
        {
            throw InvalidNativePayload("Rust 核心返回的活动预览数据无效");
        }

        if (model is null)
        {
            throw InvalidNativePayload("Rust 核心返回的活动预览数据无效");
        }

        if (model.SchemaVersion != SupportedSchemaVersion)
        {
            throw new NativeCoreException(
                IncompatibleVersionCode,
                $"Rust 核心数据版本不兼容：需要 {SupportedSchemaVersion}，实际 {model.SchemaVersion}");
        }

        if (model.AlgorithmVersion != SupportedAlgorithmVersion)
        {
            throw new NativeCoreException(
                IncompatibleVersionCode,
                $"Rust 核心算法版本不兼容：需要 {SupportedAlgorithmVersion}，实际 {model.AlgorithmVersion}");
        }

        if (model.Seed != expectedSeed
            || string.IsNullOrWhiteSpace(model.StartTimeUtc)
            || model.Laps is null
            || model.Laps.Count == 0
            || ContainsNull(model.Laps)
            || model.Samples is null
            || model.Samples.Count == 0
            || ContainsNull(model.Samples))
        {
            throw InvalidNativePayload("Rust 核心返回的活动预览数据不完整");
        }

        return model;
    }

    private static string ToUtcString(DateTime startTime)
    {
        var utc = startTime.Kind == DateTimeKind.Utc
            ? startTime
            : startTime.ToUniversalTime();
        return utc.ToString("O", CultureInfo.InvariantCulture);
    }

    private static bool ContainsNull<T>(IReadOnlyList<T> values)
        where T : class
    {
        for (var index = 0; index < values.Count; index++)
        {
            if (values[index] is null)
            {
                return true;
            }
        }

        return false;
    }

    private static bool HasFitSignature(byte[]? payload) =>
        payload is { Length: >= 12 }
        && payload[8] == (byte)'.'
        && payload[9] == (byte)'F'
        && payload[10] == (byte)'I'
        && payload[11] == (byte)'T';

    private static T InvokeNative<T>(Func<T> operation, string failureMessage)
    {
        try
        {
            return operation();
        }
        catch (NativeCoreException)
        {
            throw;
        }
        catch (Exception)
        {
            throw new NativeCoreException(InternalErrorCode, failureMessage);
        }
    }

    private static NativeCoreException InvalidNativePayload(string message) =>
        new(InternalErrorCode, message);
}

File: android/app/src/main/java/com/alphaauxiliary/fitgenerator/core/NativeCoreService.kt
package com.alphaauxiliary.fitgenerator.core

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerializationException
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString

internal class NativeCoreService(
    private val gateway: NativeCoreGateway = JniNativeCoreGateway,
    private val dispatcher: CoroutineDispatcher = Dispatchers.Default,
) {
    private val apiVersionStatus by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        readApiVersionStatus()
    }

    suspend fun preview(input: ActivityInput, seed: ULong): ActivityModelDto =
        withContext(dispatcher) {
            ensureCompatibleApi()
            val request = encodeRequest(input, seed, PREVIEW_VARIANT_INDEX)
            val payload = invokeByteOperation("无法调用 Rust 核心生成活动预览") {
                gateway.preview(request)
            }
            decodePreview(payload, seed)
        }

    suspend fun generateFit(
        input: ActivityInput,
        seed: ULong,
        variantIndex: Int,
    ): ByteArray = withContext(dispatcher) {
        ensureCompatibleApi()
        val request = encodeRequest(input, seed, variantIndex)
        val payload = invokeByteOperation("无法调用 Rust 核心生成 FIT 文件") {
            gateway.generateFit(request)
        }
        if (!hasFitSignature(payload)) {
            throw invalidNativePayload("Rust 核心返回的 FIT 数据无效")
        }
        payload
    }

    private fun readApiVersionStatus(): ApiVersionStatus =
        try {
            val actualVersion = gateway.apiVersion()
            if (actualVersion == SUPPORTED_API_VERSION) {
                ApiVersionStatus.Compatible
            } else {
                ApiVersionStatus.Failed(
                    code = NATIVE_CORE_VERSION_MISMATCH,
                    message =
                        "Rust 核心 API 版本不兼容：需要 $SUPPORTED_API_VERSION，实际 $actualVersion",
                )
            }
        } catch (_: LinkageError) {
            ApiVersionStatus.Failed(
                code = NATIVE_CORE_UNAVAILABLE,
                message = "无法加载 Rust 核心，请确认应用包含当前设备架构的原生库",
            )
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            ApiVersionStatus.Failed(
                code = NATIVE_CORE_UNAVAILABLE,
                message = "无法读取 Rust 核心 API 版本",
            )
        }

    private fun ensureCompatibleApi() {
        when (val status = apiVersionStatus) {
            ApiVersionStatus.Compatible -> Unit
            is ApiVersionStatus.Failed -> throw CoreException(status.code, status.message)
        }
    }

    private fun encodeRequest(
        input: ActivityInput,
        seed: ULong,
        variantIndex: Int,
    ): ByteArray {
        return try {
            val request = CoreRequestDto(
                schemaVersion = SUPPORTED_SCHEMA_VERSION,
                startTimeUtc = toUtcString(input.startTime),
                points = input.routePoints.map { point ->
                    GeoPointDto(lat = point.latitude, lng = point.longitude)
                },
                paceSecondsPerKm = input.paceSecondsPerKilometer,
                hrRest = input.restingHeartRate,
                hrMax = input.maximumHeartRate,
                lapCount = input.lapCount,
                variantIndex = variantIndex,
                seed = seed,
                routeMode = ROUTE_MODE,
            )
            StrictCoreJson.encodeToString(request).toByteArray(Charsets.UTF_8)
        } catch (_: SerializationException) {
            throw CoreException(INVALID_INPUT, "活动参数无法转换为 Rust 核心请求")
        } catch (_: IllegalArgumentException) {
            throw CoreException(INVALID_INPUT, "活动参数无法转换为 Rust 核心请求")
        }
    }

    private fun invokeByteOperation(
        failureMessage: String,
        operation: () -> ByteArray,
    ): ByteArray {
        val payload = try {
            operation()
        } catch (_: LinkageError) {
            throw CoreException(NATIVE_CORE_UNAVAILABLE, failureMessage)
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            throw CoreException(NATIVE_CORE_UNAVAILABLE, failureMessage)
        }

        if (payload.isEmpty()) {
            throw readCoreError()
        }
        return payload
    }

    private fun readCoreError(): CoreException {
        val rawError = try {
            gateway.lastError()
        } catch (_: LinkageError) {
            return CoreException(NATIVE_CORE_UNAVAILABLE, "无法读取 Rust 核心错误")
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            return CoreException(NATIVE_CORE_UNAVAILABLE, "无法读取 Rust 核心错误")
        }

        val error = try {
            StrictCoreJson.decodeFromString<CoreErrorDto>(rawError)
        } catch (_: SerializationException) {
            return invalidNativePayload("Rust 核心返回的错误数据无效")
        } catch (_: IllegalArgumentException) {
            return invalidNativePayload("Rust 核心返回的错误数据无效")
        }

        val code = when (error.code) {
            100 -> INVALID_INPUT
            101 -> UNSUPPORTED_SCHEMA
            102 -> RESOURCE_LIMIT
            200 -> SIMULATION_FAILED
            300 -> FIT_ENCODING_FAILED
            900 -> INTERNAL_ERROR
            else -> INTERNAL_ERROR
        }
        if (error.message.isEmpty()) {
            return invalidNativePayload("Rust 核心返回的错误消息为空")
        }
        return CoreException(code, error.message, nativeCode = error.code)
    }

    private fun decodePreview(payload: ByteArray, expectedSeed: ULong): ActivityModelDto {
        val model = try {
            StrictCoreJson.decodeFromString<ActivityModelDto>(payload.toString(Charsets.UTF_8))
        } catch (_: SerializationException) {
            throw invalidNativePayload("Rust 核心返回的活动预览数据无效")
        } catch (_: IllegalArgumentException) {
            throw invalidNativePayload("Rust 核心返回的活动预览数据无效")
        }

        if (model.schemaVersion != SUPPORTED_SCHEMA_VERSION) {
            throw CoreException(
                NATIVE_CORE_VERSION_MISMATCH,
                "Rust 核心数据版本不兼容：需要 $SUPPORTED_SCHEMA_VERSION，实际 ${model.schemaVersion}",
            )
        }
        if (model.algorithmVersion != SUPPORTED_ALGORITHM_VERSION) {
            throw CoreException(
                NATIVE_CORE_VERSION_MISMATCH,
                "Rust 核心算法版本不兼容：需要 $SUPPORTED_ALGORITHM_VERSION，实际 ${model.algorithmVersion}",
            )
        }
        if (
            model.startTimeUtc.isBlank() ||
            model.seed != expectedSeed ||
            model.laps.isEmpty() ||
            model.samples.isEmpty()
        ) {
            throw invalidNativePayload("Rust 核心返回的活动预览数据不完整")
        }
        return model
    }

    private fun toUtcString(startTime: Calendar): String {
        val localCopy = startTime.clone() as Calendar
        val formatter = SimpleDateFormat(UTC_TIMESTAMP_PATTERN, Locale.ROOT).apply {
            timeZone = UTC_TIME_ZONE
        }
        return formatter.format(Date(localCopy.timeInMillis))
    }

    private fun hasFitSignature(payload: ByteArray): Boolean =
        payload.size >= FIT_SIGNATURE_END &&
            payload[8] == '.'.code.toByte() &&
            payload[9] == 'F'.code.toByte() &&
            payload[10] == 'I'.code.toByte() &&
            payload[11] == 'T'.code.toByte()

    private fun invalidNativePayload(message: String): CoreException =
        CoreException(INTERNAL_ERROR, message, nativeCode = 900)

    private sealed interface ApiVersionStatus {
        data object Compatible : ApiVersionStatus

        data class Failed(
            val code: String,
            val message: String,
        ) : ApiVersionStatus
    }

    private companion object {
        const val SUPPORTED_API_VERSION = 1
        const val SUPPORTED_SCHEMA_VERSION = 1
        const val SUPPORTED_ALGORITHM_VERSION = 1
        const val PREVIEW_VARIANT_INDEX = 1
        const val FIT_SIGNATURE_END = 12
        const val ROUTE_MODE = "close_if_needed"
        const val UTC_TIMESTAMP_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"

        const val INVALID_INPUT = "invalid_input"
        const val UNSUPPORTED_SCHEMA = "unsupported_schema"
        const val RESOURCE_LIMIT = "resource_limit"
        const val SIMULATION_FAILED = "simulation_failed"
        const val FIT_ENCODING_FAILED = "fit_encoding_failed"
        const val INTERNAL_ERROR = "internal_error"
        const val NATIVE_CORE_UNAVAILABLE = "native_core_unavailable"
        const val NATIVE_CORE_VERSION_MISMATCH = "native_core_version_mismatch"

        val UTC_TIME_ZONE: TimeZone = TimeZone.getTimeZone("UTC")
    }
}
